invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java

```diff
@@ -48,30 +55,116 @@
+            }
+        }
+
+        return earliestRescheduleDate;
+    }
+
+    public void onSuccessCall(final LocalDate targetDate,
+                              final DefaultInvoice invoice,
+                              final List<Invoice> existingInvoices,
+                              final boolean isDryRun,
+                              final boolean isRescheduled,
+                              final CallContext callContext,
+                              final InternalTenantContext internalTenantContext) {
+        log.debug("Invoking invoice plugins onSuccessCall: targetDate='{}', isDryRun='{}', isRescheduled='{}', invoice='{}'", targetDate, isDryRun, isRescheduled, invoice);
+        onCompletionCall(true, targetDate, invoice, existingInvoices, isDryRun, isRescheduled, callContext, internalTenantContext);
+    }
+
+    public void onFailureCall(final LocalDate targetDate,
+                              final DefaultInvoice invoice,
+                              final List<Invoice> existingInvoices,
+                              final boolean isDryRun,
+                              final boolean isRescheduled,
+                              final CallContext callContext,
+                              final InternalTenantContext internalTenantContext) {
+        log.debug("Invoking invoice plugins onFailureCall: targetDate='{}', isDryRun='{}', isRescheduled='{}', invoice='{}'", targetDate, isDryRun, isRescheduled, invoice);
+        onCompletionCall(false, targetDate, invoice, existingInvoices, isDryRun, isRescheduled, callContext, internalTenantContext);
+    }
+
+    private void onCompletionCall(final boolean isSuccess,
+                                  final LocalDate targetDate,
+                                  final DefaultInvoice originalInvoice,
+                                  final List<Invoice> existingInvoices,
+                                  final boolean isDryRun,
+                                  final boolean isRescheduled,
+                                  final CallContext callContext,
+                                  final InternalTenantContext internalTenantContext) {
+        final Collection<InvoicePluginApi> invoicePlugins = getInvoicePlugins(internalTenantContext).values();
+        if (invoicePlugins.isEmpty()) {
+            return;
+        }
+
+        // We clone the original invoice so plugins don't remove/add items
+        final Invoice clonedInvoice = (Invoice) originalInvoice.clone();
+        final InvoiceContext invoiceContext = new DefaultInvoiceContext(targetDate, clonedInvoice, existingInvoices, isDryRun, isRescheduled, callContext);
+
+        for (final InvoicePluginApi invoicePlugin : invoicePlugins) {
+            if (isSuccess) {
+                invoicePlugin.onSuccessCall(invoiceContext, ImmutableList.<PluginProperty>of());
+            } else {
+                invoicePlugin.onFailureCall(invoiceContext, ImmutableList.<PluginProperty>of());
```

Comment:
If i understand correctly the `onFailureCall` scenario maps to drastic error on our end -- cannot retrieve `Account`from id, fail to write to DB, ... I suppose this will inform plugin that shit happened but we don't allow plugin to take any action. At this point, `Account` should move into `__PARKED__` mode ?

Reply:
> At this point, `Account` should move into `__PARKED__` mode ? 

Actually, back then, we had decided to park accounts only for billing errors (when `ErrorCode.UNEXPECTED_ERROR` is thrown) or when the billing system is disabled. Any other exception (`CatalogApiException`, `AccountApiException`, etc.) is ignored.

See logic in `processAccountWithLock`.

Reply:
Ah, interesting, so maybe a useful thing would be to allow plugin to decide whether account should be parked or not?